### ISSUE TITLE
Add ability to KILL BACKUP

### ISSUE THREAD

**Issue by azat** [ClickHouse/ClickHouse#52092]
**Use case**

Sometimes it may consume all the IO bandwidth, and right now there is no way to stop it, except for restarting the server.

**Describe the solution you'd like**

Actually it should be pretty easy to implement, the only thing that should be added is the `is_cancelled`/`cancellation_hook` for `copyData`

**Comment by UnamedRus** [comment]
> Sometimes it may consume all the IO bandwidth, and right now there is no way to stop it, except for restarting the server.

Starting from 23.4 it's possible to enable throttling on local disks. (IE do not use embedded backup on versions <23.4 for local disks)

**Comment by azat** [comment]
@vitlibar any progress on this one?

**Comment by azat** [comment]
Actually right now simple `KILL QUERY` by query id works (even for `ASYNC`), but this can be done even better, by checking `is_killed` flag (or similar) in `copyData` and friends.

**Comment by vitlibar** [comment]
> Actually right now simple `KILL QUERY` by query id works (even for `ASYNC`), but this can be done even better, by checking `is_killed` flag (or similar) in `copyData` and friends.

Yes, it should work, see https://github.com/ClickHouse/ClickHouse/pull/58804
It must work fine for ASYNC.
However for ON CLUSTER backups it might work not very good.

**Comment by azat** [comment]
FWIW this issue is still relevant, since it is not possible to kill the backup in the middle of the copy of a file (i.e. image you have `max_backup_bandwidth=1e6`, this can take awhile)

**Comment by groeneai** [comment]
Investigated on current master (`2a3a502d2ed0a`). This is a feature request (KILL BACKUP granularity), not a crash, so the `crash` label from the classifier is a misclassification, but `comp-backup` / @ pamarcos is a correct owner (comp-backup experts are jkartseva, hanfei1991, vitlibar, SmitaRKulkarni, pamarcos; @ vitlibar was auto-unassigned for inactivity).

**Current state**

`KILL QUERY <query_id>` already cancels a backup, including `ASYNC` (via #58804). `BackupsWorker::cancel` -> `ProcessList::sendCancelToQuery` sets `QueryStatus::is_killed`, and `writeBackupEntries` checks it before each file: `process_list_element->checkTimeLimit()` throws `QUERY_WAS_CANCELLED` (`BackupsWorker.cpp:805-806`).

So cancellation works, but only **between files**. This is exactly the remaining gap @ azat pointed out: the per-file body copy has no cancellation check, so with `max_backup_bandwidth` set, one large file blocks the kill until it finishes copying. The copy sites with no hook:
- `BackupIO_Default.cpp:35,75` -> `copyData(from, to, size)` (plain overload, no hook)
- `BackupImpl.cpp:1017,1080` -> `copyData(...)`
- `copyDataToS3File` / `copyDataToAzureBlobStorageFile` (multipart, no cancel param)

**Suggested solution**

`copyData` already has the needed overloads (`IO/copyData.h:26-27`): `copyData(from, to, [, bytes], std::function<void()> cancellation_hook)`. The fix is to thread the `QueryStatusPtr` (already available as `process_list_element` in `BackupsWorker`) down into the writer copy path and pass:

```cpp
auto cancellation_hook = [ple = process_list_element]{ if (ple) ple->checkTimeLimit(); };
```

`checkTimeLimit()` throws `QUERY_WAS_CANCELLED` when killed, so the hook doubles as both KILL and `max_execution_time` enforcement mid-copy. Scope:
1. `BackupIO_Default::copyDataToFile` / `BackupImpl` copy calls: switch to the `cancellation_hook` overload.
2. `copyDataToS3File` / `copyDataToAzureBlobStorageFile`: add an optional `cancellation_hook`, checked between multipart uploads (checking inside a single in-flight part isn't feasible, but part granularity is enough to bound latency).
3. Plumb `process_list_element` through `IBackupWriter::copyDataToFile` (or store it on the writer) so the hook is available at the copy site.

Local-disk copies are already interruptible at part/buffer granularity via `copyData`'s `is_cancelled` loop once the hook is passed. Note (per @ vitlibar earlier): `ON CLUSTER` backups still cancel less cleanly, since a KILL on the initiator doesn't reliably propagate the cancel to secondary nodes' internal backup queries; that would need coordination-level cancel propagation and is worth tracking separately.

I can put up a PR for the single-node mid-file cancellation if that's wanted.
